BitFaster.Caching/Lru/TemplateConcurrentLru.cs

```diff
@@ -310,12 +310,101 @@ public void Clear()
             // and queues will remain in a consistent state.
             for (int i = 0; i < keys.Count; i++)
             {
-                CycleHotUnchecked();
-                CycleWarmUnchecked();
-                CycleColdUnchecked();
+                CycleHotUnchecked(ItemRemovedReason.Clear);
+                CycleWarmUnchecked(ItemRemovedReason.Clear);
+                CycleColdUnchecked(ItemRemovedReason.Clear);
             }
         }
 
+        // Removes d discardable items per IItemPolicy.ShouldDiscard(), then itemCount-d items in LRU order, if any.
+        // Note: this resets accessed status of items.
+        public int Trim(int itemCount)
+        {
+            int capacity = this.coldCapacity + this.warmCapacity + this.hotCapacity;
+
+            if (itemCount < 0 || itemCount > capacity)
+            { 
+                throw new ArgumentOutOfRangeException(nameof(itemCount), "itemCount must be greater than or equal to zero, and less than the capacity of the cache.");
+            }
+
+            // clamp itemCount to number of items actually in the cache - it shouldn't be larger
+            itemCount = Math.Min(itemCount, this.HotCount + this.WarmCount + this.ColdCount);
+
+            int attempts = 0;
+            int itemsRemoved = 0;
+
+            // first scan each queue for discardable items and remove them immediately. Note this can remove > itemCount items.
+            void RemoveDiscardableItems(ConcurrentQueue<I> q, ref int queueCounter)
+            {
+                int localCount = queueCounter;
+
+                for (int i = 0; i < localCount; i++)
+                {
+                    if (q.TryDequeue(out var item))
+                    {
+                        if (this.itemPolicy.ShouldDiscard(item))
+                        {
+                            Interlocked.Decrement(ref queueCounter);
+                            this.Move(item, ItemDestination.Remove, ItemRemovedReason.Trim);
+                            itemsRemoved++;
+                        }
+                        else
+                        {
+                            q.Enqueue(item);
+                        }
+                    }
+                }
+            }
+
+            RemoveDiscardableItems(coldQueue, ref this.coldCount);
+            RemoveDiscardableItems(warmQueue, ref this.warmCount);
+            RemoveDiscardableItems(hotQueue, ref this.hotCount);
+            
+            // If there is no activity while this loop is running, it will simply remove itemCount items.
+            // If items are constantly added while running, size will be approx. same and eventually
+            // attempts will be >= capacity terminating the loop.
+            while (itemsRemoved < itemCount && attempts < capacity)

```

Comment:
Limitation here is that Trim can remove an item while another thread adds an item.

Initial and end count will be the same, so even though an item was removed (and an event can fire with reason==Trimmed) we will not capture that and could return 0 as number of elements trimmed.